File: clients/Scala/src/main/scala/model/Player.scala
package model

import util.StreamUtil

case class Player(id: Int, score: Int, resource: Int) {
    def writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, id)
        StreamUtil.writeInt(stream, score)
        StreamUtil.writeInt(stream, resource)
    }
}
object Player {
    def readFrom(stream: java.io.InputStream): Player = Player(
        StreamUtil.readInt(stream)
        ,
        StreamUtil.readInt(stream)
        ,
        StreamUtil.readInt(stream)
        )
}

File: clients/Cpp/model/Player.cpp
#include "Player.hpp"

Player::Player() { }
Player::Player(int id, int score, int resource) : id(id), score(score), resource(resource) { }
Player Player::readFrom(InputStream& stream) {
    Player result;
    result.id = stream.readInt();
    result.score = stream.readInt();
    result.resource = stream.readInt();
    return result;
}
void Player::writeTo(OutputStream& stream) const {
    stream.write(id);
    stream.write(score);
    stream.write(resource);
}
bool Player::operator ==(const Player& other) const {
    return id == other.id && score == other.score && resource == other.resource;
}
size_t std::hash<Player>::operator ()(const Player& value) const {
    size_t result = 0;
    result ^= std::hash<int>{}(value.id) + 0x9e3779b9 + (result<<6) + (result>>2);
    result ^= std::hash<int>{}(value.score) + 0x9e3779b9 + (result<<6) + (result>>2);
    result ^= std::hash<int>{}(value.resource) + 0x9e3779b9 + (result<<6) + (result>>2);
    return result;
}
